### PR TITLE
Fix ruff error PTH124

### PR DESCRIPTION
**Pull request by draghuram**
### Description

This pull request is to address ruff errors PTH124 ~and PLW0602~. I ran all the tests and there were no failures.

- [ ] By checking this box, the PR author has requested that maintainers do **NOT** use the "Squash and Merge" button. Maintainers should respect this when possible; however, the final decision is at the discretion of the maintainer that merges the PR.

### CODE REVIEW

File: astropy/utils/data.py
Comment (pllim) [inline astropy/utils/data.py:1]: Wait, what?

**Review by pllim — CHANGES_REQUESTED** [review]
Each code should be handled separately. So this should have been 2 PRs.

Also the change to `utils/data.py` makes no sense to me.

File: astropy/utils/tests/test_data.py
Comment (eerovaher) [inline astropy/utils/tests/test_data.py:1607]: The problem with this change is that now `test_get_fileobj_localpath()` is exactly the same as `test_get_fileobj_pathlib()`, which is the very next test in this file. Such duplication should be avoided.

[`py.path`](https://py.readthedocs.io/en/latest/path.html) provided a high-level API for filesystem paths, but it became obsolete when Python 3.4 introduced the built-in [`pathlib`](https://docs.python.org/3/library/pathlib.html) module. The question here is whether `astropy` should still explicitly check that `get_readable_fileobj()` works with the legacy `py.path`, in which case there should be a [`noqa` comment](https://beta.ruff.rs/docs/tutorial/#ignoring-errors) on this line, or whether we no longer care about `py.path` enough to have this test.

**Review by eerovaher — DISMISSED** [review]
Pull requests that address Ruff issues are very welcome, and pull requests that address multiple Ruff rules are also welcome as long as the total number of edited lines is small (in this case it clearly is). But it would be better if each rule is addressed in a separate commit.

Rule [PLW0602](https://beta.ruff.rs/docs/rules/global-variable-not-assigned/) has been correctly dealt with, but [PTH124](https://beta.ruff.rs/docs/rules/py-path/) is much trickier and we need input from `utils` maintainers (@pllim, @bsipocz, @eteq) to proceed.

File: astropy/utils/data.py
Comment (eerovaher) [reply astropy/utils/data.py:1]: Global variables have to be explicitly declared as global only if they are assigned to inside the function. In this case nothing is being assigned to `_tempfilestodel`, so there's no need to use `global` and it's possible to shorten the code.

Note that if Python would require all global variables to be explicitly declared then a simple `hello_world()` would have to be implemented as
```python
def hello_world():
    global print
    print("Hello, World!")
```

Comment (pllim) [reply astropy/utils/data.py:1]: You used to have to declare it as global if you want to modify it inside a local function. Has that changed?

Comment (draghuram) [reply astropy/utils/data.py:1]: That hasn't changed but in this case, the variable was not being modified. Calling a method on the object doesn''t require global declaration.

**Comment by draghuram** [comment]
Thanks for reviewing it. I combined two rules as the number of changes is small. But I can keep them as separate commits in future if that makes more sense.

**Comment by draghuram** [comment]
> Pull requests that address Ruff issues are very welcome, and pull requests that address multiple Ruff rules are also welcome as long as the total number of edited lines is small (in this case it clearly is). But it would be better if each rule is addressed in a separate commit.
> 
> Rule [PLW0602](https://beta.ruff.rs/docs/rules/global-variable-not-assigned/) has been correctly dealt with, but [PTH124](https://beta.ruff.rs/docs/rules/py-path/) is much trickier and we need input from `utils` maintainers (@pllim, @bsipocz, @eteq) to proceed.

@eerovaher, Since PLW0602 is properly handled, I can break that change into a separate PR while we wait for feedback regarding the other one. Please let me know your thoughts.

**Comment by eerovaher** [comment]
If a pull request addresses unrelated issues then it can easily happen (like is indeed happening here) that one of the issues is resolved quickly but the other one is not, which can greatly delay merging the uncontroversial patch. Furthermore, the pull request comments can contain two unrelated discussions, which is not helpful. However, spreading discussion about one issue over multiple pull requests is not helpful either. In this case the mixing of unrelated issues has already occurred, splitting related discussion over multiple pull requests can still be avoided.

I do think that addressing each rule in a separate commit is a good idea.

**Comment by draghuram** [comment]
Thanks. Perhaps, we should wait for some more time to hear from utils maintainers (@pllim, @bsipocz, @eteq) before I split the PR.

**Comment by pllim** [comment]
Please do split the PR.

I'll have to remind myself how global variables work in Python again before I come back to it.

**Comment by draghuram** [comment]
Sure, will do.

**Comment by draghuram** [comment]
I split the PR as you can see. I kept this PR for PTH124 and created a new one for PLW0602.

File: astropy/utils/tests/test_data.py
Comment (pllim) [reply astropy/utils/tests/test_data.py:1607]: > we no longer care about py.path enough to have this test

I am leaning towards not caring.

Comment (draghuram) [reply astropy/utils/tests/test_data.py:1607]: Ok. Should I go ahead and remove the test all together? The PR would pretty be removing "PTH124" from the ignore list.

Comment (pllim) [reply astropy/utils/tests/test_data.py:1607]: I think that was what @eerovaher was eluding to but let wait to see what he says. Thanks!

Comment (eerovaher) [reply astropy/utils/tests/test_data.py:1607]: As I already said there is no point in replacing `py.path.local` with `pathlib.Path` because the latter is already being tested in `test_get_fileobj_pathlib()`. So if testing with `py.path` is no longer considered important then `test_get_fileobj_localpath()` serves no purpose and should indeed be deleted.

Comment (nstarman) [reply astropy/utils/tests/test_data.py:1607]: I agree. Let's rm the test. Thanks!

**Review by nstarman — APPROVED** [review]
LGTM! Let's let @eerovaher & @pllim  take a look as well.

**Review by pllim — APPROVED** [review]
LGTM. Thanks!

**Pull request merged**
